### ISSUE TITLE
Categorical entropy of logits is inconsistent with probs

### ISSUE THREAD

**Issue by aray-openai** [pytorch/pytorch#40553]
## 🐛 Bug

When using the `Categorical` distribution, having 0-probability classes results in correct entropy calculations, while using the same distribution as logits (which includes `-inf`) does not.

## To Reproduce

```
In [1]: import torch; from torch.distributions import Categorical

In [2]: logits = torch.tensor([0, 0, -float('inf')])

In [3]: Categorical(logits=logits).entropy()
Out[3]: tensor(nan)

In [4]: Categorical(probs=logits.softmax(-1)).entropy()
Out[4]: tensor(0.6931)
```

## Expected behavior

Entropy should be the same in both cases.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5.0
 - OS (e.g., Linux):  Mac OSX 10.14.6
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information:

## Additional context

I hope this is a simple fix?

cc @ezyang @gchanan @zou3519 @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

**Comment by alicanb** [comment]
Workaround is to use a large negative value instead of -inf
```
logits = torch.tensor([0, 0, -1e16])
d = Categorical(logits=logits)
d.entropy() # tensor(0.6931)
```

one possible idea is clamp logits between `torch.finfo(logits.dtype).min` and `torch.finfo(logits.dtype).max` during initialization.

another is to change definition of https://github.com/pytorch/pytorch/blob/c362138f4380c11ddeb07d7e7e34d75300091597/torch/distributions/constraints.py#L158 to exclude infs, so that `Categorical(logits=torch.tensor([0, 0, -float('inf')]), validate_args=True)`

wdyt?

**Comment by mattip** [comment]
There is no definition of `logits` in the [Categorical](https://pytorch.org/docs/master/distributions.html#categorical) documentation. It does refer to [Multinomial](https://pytorch.org/docs/master/distributions.html#multinomial) which states 
> `logit`: event log probabilities"

and 
> `prob`: event probabilities". 

It also has a note 

> 'probs' must be non-negative, finite and have a non-zero sum, and it will be normalized to sum to 1. 

So it would seem `-inf` is a valid `logit` value (should result in a `prob` of 0, which is allowed).

@alicanb it seems your comment was truncated, did you mean this?

> `Categorical(logits=torch.tensor([0, 0, -float('inf')]), validate_args=True)` should fail

**Comment by mattip** [comment]
With gh-41002, the test at the top of the issue passes.
